Dockerfile

```diff
@@ -1,26 +1,48 @@
-# Build from local source
+# Build from local source and run tests
 #   docker build --tag stencila/encoda .
-# Run tests
-#   docker run stencila/encoda
-# Or to test out interactively at the terminal
-#   docker run -it stencila/encoda bash
+#   docker run -it --init --rm --cap-add=SYS_ADMIN stencila/encoda
 
-FROM node
+# Or instead, the above is as a npm script:
+#   npm test:docker
+#
+# To test out interactively at the terminal
+#   docker run -it --init --rm --cap-add=SYS_ADMIN stencila/encoda bash
+#
+# The `--init` and `--cap-add` options to `run` are necessary for running Puppeteer
+# (see below).
 
-RUN mkdir -p /opt/encoda
-WORKDIR /opt/encoda
+FROM node:12
 
-# Separate copy and Docker layer for `package.json` to prevent
-# unecessary reinstall of `node_modules` when `src` changes
-COPY package.json package.json
-RUN npm install
+# Installs for getting Puppeteer to run in Docker
+# See the following for more, including recommended `docker run` args
+#  https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker
+#  https://github.com/puppeteer/puppeteer/issues/3451#issuecomment-523961368
+
+# Install necessary libs to make the bundled version of Chromium that Puppeteer installs, work.
+# From https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker
+RUN wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
+    && sh -c 'echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google.list' \
+    && apt-get update \
+    && apt-get install -y google-chrome-unstable fonts-ipafont-gothic fonts-wqy-zenhei fonts-thai-tlwg fonts-kacst fonts-freefont-ttf \
+      --no-install-recommends \
+    && rm -rf /var/lib/apt/lists/*
 
-# Copy over everything else
-COPY . .
+# Enable user namespace cloning so that Chromium can create a sandbox
+RUN echo 'kernel.unprivileged_userns_clone=1' > /etc/sysctl.d/userns.conf
 
-# Run as guest user
-RUN useradd --create-home --home-dir /home/guest guest
-RUN chown -R guest /opt/encoda
-USER guest
+# Install and run as non-root user, because it is generally good practice,
+# and in this case necessary to be able to run Chromium without the
+# unsafe `--no-sandbox` option.
+RUN useradd --create-home encoda
+WORKDIR /home/encoda
+USER encoda
+
+# Copy over everything and install.
+# Do not do the dusual Node package / Docker optimization of copying over
+# `package.json` and installing in a seperate layer because that is not possible given
+# how our `install.js` script works.
+COPY --chown=encoda:encoda . .
```

Comment:
the separate `chown -R` was a performance killer, so this helps.
Though is really any file changes in the root a trigger for a new `npm install`? For example, even a change to `README.md` (doesn't seem `.dockerignore`d) or a new line at the end of the `Dockerfile` (can't be `.dockerignore`d) would result in a new execution. Could list out explicitly the folders with various `COPY` instructions?

Reply:
Yes, good point. I have in the past been bitten by not including all necessary files (IIRC) so have formed a habit of `COPY . .`, which is indeed sub-optimal. Well revise. Thanks!